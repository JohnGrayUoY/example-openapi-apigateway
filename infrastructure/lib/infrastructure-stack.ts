import { Stack, type StackProps } from 'aws-cdk-lib';
import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import type { Construct } from 'constructs';
import { ApiDefinition, Deployment, SpecRestApi, Stage } from 'aws-cdk-lib/aws-apigateway';
import { Effect, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { type Environment } from '../types/parseEnvironment';
import { createWorkordersAPIDocument } from '../../src/openAPI/createWorkordersAPIDocument';

export class InfrastructureStack extends Stack {
    constructor(scope: Construct, id: string, env: Environment, props?: StackProps) {
        super(scope, id, props);

        const getWorkordersLambda = new NodejsFunction(this, 'Get-Workorders-Lambda', {
            runtime: Runtime.NODEJS_22_X,
            architecture: Architecture.ARM_64,
            functionName: `Get-Workorders-Lambda`,
            entry: 'src/lambda/index.ts',
            handler: 'handler',
            environment: {
                status: env.STATUS,
            },
        });

        const apiRole = new Role(this, 'API-Role', {
            roleName: 'API-Role',
            assumedBy: new ServicePrincipal('apigateway.amazonaws.com'),
        });

        apiRole.addToPolicy(
            new PolicyStatement({
                effect: Effect.ALLOW,
                resources: [getWorkordersLambda.functionArn],
                actions: ['lambda:InvokeFunction'],
            })
        );

        const openAPIDocument = createWorkordersAPIDocument({
            getWorkordersLambdaFunctionName: getWorkordersLambda.functionName,
            apiRoleArn: apiRole.roleArn,
        });

        const api = new SpecRestApi(this, 'FinanceAPI', {
            apiDefinition: ApiDefinition.fromInline(openAPIDocument),
        });

        const deployment = new Deployment(this, 'FinanceAPIDeployment', { api });

        const [_devStage, _stagingStage, productionStage] = ['dev', 'staging', 'production'].map(
            (stage) => new Stage(this, `${stage}_stage`, { deployment, stageName: stage })
        );

        let deploymentStage: Stage;

        switch (env.STATUS) {
            case 'dev': {
                deploymentStage = _devStage!;
                break;
            }

            case 'staging': {
                deploymentStage = _stagingStage!;
                break;
            }

            default: {
                deploymentStage = productionStage!;
                break;
            }
        }

        api.deploymentStage = deploymentStage;
    }
}
