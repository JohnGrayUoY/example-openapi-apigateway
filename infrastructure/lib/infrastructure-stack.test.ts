import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { InfrastructureStack } from './infrastructure-stack';

describe('InfrastructureStack', () => {
    it('should compile', () => {
        const app = new App();
        const stack = new InfrastructureStack(app, 'infrastructure-stack', { STATUS: 'test' });
        expect(Template.fromStack(stack)).toMatchSnapshot();
    });
});
