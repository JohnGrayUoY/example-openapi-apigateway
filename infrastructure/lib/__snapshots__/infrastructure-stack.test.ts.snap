// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InfrastructureStack > should compile 1`] = `
{
  "Outputs": {
    "FinanceAPIEndpoint48E49EBE": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "FinanceAPI24BBA5D4",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "FinanceAPIDeploymentStageprod417D1B6E",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "APIRole89BF3F08": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "API-Role",
      },
      "Type": "AWS::IAM::Role",
    },
    "APIRoleDefaultPolicy4E892C15": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "GetWorkordersLambda403E65B1",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "APIRoleDefaultPolicy4E892C15",
        "Roles": [
          {
            "Ref": "APIRole89BF3F08",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FinanceAPI24BBA5D4": {
      "Properties": {
        "Body": {
          "components": {
            "x-amazon-apigateway-policy": {
              "Statement": [
                {
                  "Action": "execute-api:Invoke",
                  "Effect": "Allow",
                  "Principal": {
                    "AWS": "*",
                  },
                  "Resource": [
                    "execute-api:/*",
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
          },
          "info": {
            "title": "Workorders API",
            "version": "0.1.0",
          },
          "openapi": "3.0.3",
          "paths": {
            "/workorders": {
              "get": {
                "description": "Get workorders",
                "responses": {
                  "200": {
                    "content": {
                      "application/json": {
                        "schema": {
                          "properties": {
                            "workorders": {
                              "items": {
                                "properties": {
                                  "id": {
                                    "description": "The id of the workorder.",
                                    "example": "CorpoServo000",
                                    "type": "string",
                                  },
                                },
                                "required": [
                                  "id",
                                ],
                                "type": "object",
                              },
                              "type": "array",
                            },
                          },
                          "required": [
                            "workorders",
                          ],
                          "type": "object",
                        },
                      },
                    },
                    "description": "200 OK",
                  },
                },
                "x-amazon-apigateway-auth": {
                  "type": "NONE",
                },
                "x-amazon-apigateway-integration": {
                  "credentials": {
                    "Fn::GetAtt": [
                      "APIRole89BF3F08",
                      "Arn",
                    ],
                  },
                  "httpMethod": "POST",
                  "type": "aws_proxy",
                  "uri": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:\${AWS::Partition}:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/arn:\${AWS::Partition}:lambda:\${AWS::Region}:\${AWS::AccountId}:function:",
                        {
                          "Ref": "GetWorkordersLambda403E65B1",
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
            },
          },
        },
        "Name": "FinanceAPI",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "FinanceAPIAccountA09C8324": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "FinanceAPI24BBA5D4",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "FinanceAPICloudWatchRoleBF59266B",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "FinanceAPICloudWatchRoleBF59266B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "FinanceAPIDeployment407C39B8b65a58e0fa9c70fb3141b11df70f8464": {
      "Properties": {
        "RestApiId": {
          "Ref": "FinanceAPI24BBA5D4",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "FinanceAPIDeployment6EDA9FBBb65a58e0fa9c70fb3141b11df70f8464": {
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "FinanceAPI24BBA5D4",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "FinanceAPIDeploymentStageprod417D1B6E": {
      "DependsOn": [
        "FinanceAPIAccountA09C8324",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "FinanceAPIDeployment6EDA9FBBb65a58e0fa9c70fb3141b11df70f8464",
        },
        "RestApiId": {
          "Ref": "FinanceAPI24BBA5D4",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "GetWorkordersLambda403E65B1": {
      "DependsOn": [
        "GetWorkordersLambdaServiceRole2ED718E0",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": "[CODE CONFIG OMITTED]",
        "Environment": {
          "Variables": {
            "status": "test",
          },
        },
        "FunctionName": "Get-Workorders-Lambda",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "GetWorkordersLambdaServiceRole2ED718E0",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GetWorkordersLambdaServiceRole2ED718E0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "devstageAD791BF6": {
      "DependsOn": [
        "FinanceAPIAccountA09C8324",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "FinanceAPIDeployment407C39B8b65a58e0fa9c70fb3141b11df70f8464",
        },
        "RestApiId": {
          "Ref": "FinanceAPI24BBA5D4",
        },
        "StageName": "dev",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "productionstageD3A4C83E": {
      "DependsOn": [
        "FinanceAPIAccountA09C8324",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "FinanceAPIDeployment407C39B8b65a58e0fa9c70fb3141b11df70f8464",
        },
        "RestApiId": {
          "Ref": "FinanceAPI24BBA5D4",
        },
        "StageName": "production",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "stagingstageBE297314": {
      "DependsOn": [
        "FinanceAPIAccountA09C8324",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "FinanceAPIDeployment407C39B8b65a58e0fa9c70fb3141b11df70f8464",
        },
        "RestApiId": {
          "Ref": "FinanceAPI24BBA5D4",
        },
        "StageName": "staging",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
